kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java

```diff
@@ -112,11 +106,11 @@ public ConsumerInstanceId getId() {
 
 
   public void startRead() {
-    lock.readLock().lock();
+    lock.lock();
```

Comment:
My main concern is that we hold onto this lock even when the task is not scheduled. This is probably why we tried to use the read lock before even though it didn't work. In the worst case, this can cause deadlock if a task which depends on this lock is scheduled ahead of the one that is holding it. I'm wondering if it would be better to hold the lock only during the `partialRead` operation. The main drawback of doing so as far as I can tell is that certain "write" operations may interfere with a read which is actively polling for data. For example, if there is a concurrent `seek`, then the fetch may include data both from before and after the `seek`. 

One way to solve this would be to have a queue of pending operations for each consumer instance and only submit them to the executor one at a time. I think this can be left for future work though. It's more important to address the problems with the current threading.

Reply:
I see what you mean.
I am not very knowledgeable on the thread scheduling internals. In the deadlock case, wouldn't a task that is stuck on a lock eventually be context-switched out of? A deadlock happens when both threads want something the other one is holding whereas in this case, only one thread wants something the other is holding.  The original thread which has the lock would be free to proceed?

In regards to a concurrent `seek`, that behavior seems to be described in the javadoc of the method in Kafka itself:
```java
    /**
     * Overrides the fetch offsets that the consumer will use on the next {@link #poll(long) poll(timeout)}. If this API
     * is invoked for the same partition more than once, the latest offset will be used on the next poll(). **Note that
     * you may lose data if this API is arbitrarily used in the middle of consumption, to reset the fetch offsets**
...
    public void seek(TopicPartition partition, long offset) {
```
so maybe it is expected behavior?

My concern is what a concurrent fetch would do. I imagine there would be a possibility where we return non-consecutive offsets from both requests, e.g
```
request A
messages with offset 10-20. 40-60

request B
messages with offset 0-10. 20-40
```
because that is what they both got respectively from the `poll()` calls.

Reply:
I agree it is a concern, but I'm not sure the serialized read behavior gives you what you want either. If there are concurrent requests, there's a good chance one of them is a retry, so this behavior would ensure that the user loses data because the response to the first request will never be received. 

To be honest, I don't see how we can have good semantics if the read request doesn't take an expected fetch offset. I think it's more important at the moment to ensure that reads can always make progress even if the read semantics are not ideal.